File: karakoram/kb/kb0000/src/main/java/com/anonymouscorgi/karakoram/kb0000/LeetCode034FindFirstandLastPositionofElementinSortedArray.java
package com.anonymouscorgi.karakoram.kb0000;

/**
 * Given an array of integers nums sorted in non-decreasing order, find the starting and ending
 * position of a given target value.
 *
 * If target is not found in the array, return [-1, -1].
 *
 * You must write an algorithm with O(log n) runtime complexity.
 *
 *
 * Example 1: Input: nums = [5,7,7,8,8,10], target = 8 Output: [3,4]
 *
 * Example 2: Input: nums = [5,7,7,8,8,10], target = 6 Output: [-1,-1]
 *
 * Example 3: Input: nums = [], target = 0 Output: [-1,-1]
 *
 *
 * Constraints:
 *
 *
 * 0 <= nums.length <= 10⁵ -10⁹ <= nums[i] <= 10⁹ nums is a non-decreasing array. -10⁹ <= target <=
 * 10⁹
 */
interface LeetCode034FindFirstandLastPositionofElementinSortedArray {

  int[] searchRange(int[] nums, int target);

  LeetCode034FindFirstandLastPositionofElementinSortedArray Method =
      new LeetCode034FindFirstandLastPositionofElementinSortedArray() {

        @Override
        public int[] searchRange(int[] nums, int target) {
          if (nums == null || nums.length == 0) {
            return new int[]{-1, -1};
          }

          int[] res = {-1, -1};
          int start = 0;
          int end = nums.length - 1;
          while (start < end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] < target) {
              start = mid + 1;
            } else {
              end = mid;
            }
          }

          if (nums[end] != target) {
            return res;
          }
          res[0] = end;

          end = nums.length - 1;
          while (start < end) {
            int mid = start + (end - start + 1) / 2;
            if (nums[mid] > target) {
              end = mid - 1;
            } else {
              start = mid;
            }
          }
          res[1] = start;
          return res;
        }
      };
}

File: karakoram/kb/kb0000/src/main/kotlin/com/anonymouscorgi/karakoram/kb0000/LeetCode034FindFirstandLastPositionofElementinSortedArray.kt
package com.anonymouscorgi.karakoram.kb0000

internal interface LeetCode034FindFirstandLastPositionofElementinSortedArray {

  fun searchRange(nums: IntArray, target: Int): IntArray

  object METHOD : LeetCode034FindFirstandLastPositionofElementinSortedArray {

    override fun searchRange(nums: IntArray, target: Int): IntArray {
      if (nums.isEmpty()) return intArrayOf(-1, -1)

      var leftIndex = 0
      var rightIndex = nums.lastIndex
      while (leftIndex < rightIndex) {
        val midIndex = leftIndex + (rightIndex - leftIndex) / 2;
        when {
          nums[midIndex] < target -> leftIndex = midIndex + 1
          else -> rightIndex = midIndex
        }
      }
      val targetStart = rightIndex
      if (nums[targetStart] != target) {
        return intArrayOf(-1, -1)
      }

      leftIndex = maxOf(targetStart, 0)
      rightIndex = nums.lastIndex
      while (leftIndex < rightIndex) {
        val midIndex = leftIndex + (rightIndex - leftIndex + 1) / 2;
        when {
          nums[midIndex] > target -> rightIndex = midIndex - 1
          else -> leftIndex = midIndex
        }
      }
      val targetEnd = leftIndex

      return intArrayOf(targetStart, targetEnd)
    }
  }

  object Method2 : LeetCode034FindFirstandLastPositionofElementinSortedArray {
    override fun searchRange(nums: IntArray, target: Int): IntArray {
      if (nums.isEmpty()) {
        return intArrayOf(-1, -1)
      }

      val res = intArrayOf(-1, -1)
      var start = 0
      var end = nums.size - 1
      while (start < end) {
        val mid = start + (end - start) / 2
        if (nums[mid] < target) {
          start = mid + 1
        } else {
          end = mid
        }
      }

      if (nums[end] != target) {
        return res
      }
      res[0] = end

      end = nums.size - 1
      while (start < end) {
        val mid = start + (end - start + 1) / 2
        if (nums[mid] > target) {
          end = mid - 1
        } else {
          start = mid
        }
      }
      res[1] = start
      return res
    }
  }
}